### PR TITLE
Graduate the admission and admissionregistration (webhook part) API to v1beta1

### PR DESCRIPTION
**Pull request by caesarxuchao**
ref: kubernetes/features#492

Most changes are mechanical. Please take a look at the commit message to see if the commit is worth reviewing.

```release-note
Action required:
The `admission/v1alpha1` API has graduated to `v1beta1`. Please delete your existing admission webhooks before upgrading the cluster, and update them to use the latest API, because the API has backwards incompatible changes.
The admission webhook configurations API, which is part of the `admissionregistration` API, has graduated to `v1beta1`. Please delete your existing configurations before upgrading the cluster, and update your configuration files to use the latest API.
```

### CODE REVIEW

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -560,6 +561,7 @@ func BuildStorageFactory(s *options.ServerRunOptions) (*serverstorage.DefaultSto
 		[]schema.GroupVersionResource{
 			batch.Resource("cronjobs").WithVersion("v1beta1"),
 			storage.Resource("volumeattachments").WithVersion("v1alpha1"),
+			admissionregistration.Resource("initializerconfigurations").WithVersion("v1alpha1"),
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: What exactly are the things in this list?

**Comment by lavalamp** [comment]
lgtm, will give others a chance to see before tagging it.

**Comment by lavalamp** [comment]
Should the action required note advise deleting any webhooks configured under the v1alpha1 api before upgrading?

**Comment by liggitt** [comment]
> Should the action required note advise deleting any webhooks configured under the v1alpha1 api before upgrading?

Yes, I think the schema changed incompatibly in the alpha object types

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (caesarxuchao) [reply cmd/kube-apiserver/app/server.go]: These are the objects that need to be serialized to a different version before storing to etcd than the other objects in the same group.

**Comment by caesarxuchao** [comment]
Updated the release note. Failing tests should have been fixed.

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (lavalamp) [reply cmd/kube-apiserver/app/server.go]: Since it looks like there's still a test to fix, can you add a comment to that effect?

**Comment by lavalamp** [comment]
/approve

lgtm when tests pass.

**Comment by caesarxuchao** [comment]
The cross build failure is https://github.com/kubernetes/kubernetes/issues/56002

**Comment by caesarxuchao** [comment]
Applying the lgtm label based on https://github.com/kubernetes/kubernetes/pull/56004#issuecomment-345777909.

**Comment by deads2k** [comment]
/hold

I'd like 24 hours to double check against requirements for beta.

**Comment by caesarxuchao** [comment]
cc @cheftako

File: test/e2e/apimachinery/webhook.go [hunk 2]
```diff
@@ -87,7 +87,7 @@ var _ = SIGDescribe("AdmissionWebhook", func() {
 		framework.SkipUnlessServerVersionGTE(serverWebhookVersion, f.ClientSet.Discovery())
 		framework.SkipUnlessProviderIs("gce", "gke", "local")
 
-		_, err := f.ClientSet.AdmissionregistrationV1alpha1().ValidatingWebhookConfigurations().List(metav1.ListOptions{})
+		_, err := f.ClientSet.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().List(metav1.ListOptions{})
 		if errors.IsNotFound(err) {
 			framework.Skipf("dynamic configuration of webhooks requires the alpha admissionregistration.k8s.io group to be enabled")
```
Comment (roycaihw) [inline test/e2e/apimachinery/webhook.go]: just a nit: the error message "requires the alpha admissionregistration.k8s.io group" can be changed to "beta"

cc @yliaog

Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: Thanks. Will do that in a followup.

**Comment by sttts** [comment]
In https://github.com/kubernetes/kubernetes/pull/55282/commits/9ddea83a2ce0937cf0fc8f8c35614bb18e74cfad#diff-4fbc9e9454488ac9d5b46d8fda0604f3R386 the storage path for the webhook config was changed. This means we do not preserve the alpha objects. This means we have to change the release notes, i.e. that old webhook configs are lost.

**Comment by deads2k** [comment]
> In 9ddea83#diff-4fbc9e9454488ac9d5b46d8fda0604f3R386 the storage path for the webhook config was changed. This means we do not preserve the alpha objects. This means we have to change the release notes, i.e. that old webhook configs are lost.

Yeah.  I think Jordan noted that.  It's fine given the status of the feature.

**Comment by deads2k** [comment]
> Checking against the work items listed in the proposal, I think the only thing left is "Document recommended flag order for admission plugins", which is being worked on in kubernetes/website#6368.

I've had a chance to look through and I agree.  I've also been able to sketch out a fairly complete story for someone trying to use it with what we have, so I think we can do this.

We do have one outstanding problem that is going to impact usage: https://github.com/kubernetes/kubernetes/issues/55435, but I think that's a problem with the admission plugin code itself, not the mechanisms.

/hold cancel

**Comment by caesarxuchao** [comment]
Thanks for the reviews.

File: test/e2e/apimachinery/webhook.go (hunk 2 above)
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: Done

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (caesarxuchao) [reply cmd/kube-apiserver/app/server.go]: Done.

**Pull request merged**